src/main/java/com/gluonhq/richtext/RichTextAreaSkin.java

```diff
@@ -138,18 +142,51 @@ public void dispose() {
     //TODO Need more optimal way of rendering text fragments.
     //  For now rebuilding the whole text flow
     private void refreshTextFlow() {
-        var fragments = new ArrayList<Text>();
-        viewModel.walkFragments( (text, decoration) -> fragments.add( buildText(text, decoration)));
-        textFlow.getChildren().setAll(fragments);
+        fontCacheEvictionTimer.pause();
+        try {
+            var fragments = new ArrayList<Text>();
+            viewModel.walkFragments((text, decoration) -> fragments.add(buildText(text, decoration)));
+            textFlow.getChildren().setAll(fragments);
+        } finally {
+            fontCacheEvictionTimer.start();
+        }
     }
 
     private Text buildText(String content, TextDecoration decoration ) {
-        Text text = new Text(content);
-        Optional.ofNullable(decoration.getForeground()).ifPresent(text::setFill); // has to be fill for font to render properly
-        Optional.ofNullable(decoration.getFont()).ifPresentOrElse(text::setFont, () -> text.setFont(TextDecoration.DEFAULT.getFont()) );
+        Objects.requireNonNull(decoration);
+        Text text = new Text(Objects.requireNonNull(content));
+        text.setFill(decoration.getForeground());
+
+        // Cashing fonts, assuming their reuse, especially for default one
+        int hash = Objects.hash(
+                decoration.getFontFamily(),
+                decoration.getFontWeight(),
+                decoration.getFontPosture(),
+                decoration.getFontSize());
+
+        Font font = fontCache.computeIfAbsent( hash,
+            h -> Font.font(
+                    decoration.getFontFamily(),
+                    decoration.getFontWeight(),
+                    decoration.getFontPosture(),
+                    decoration.getFontSize()));
+
+        text.setFont(font);
         return text;
     }
 
+    private void evictUnusedFonts() {
+        List<Font> usedFonts =  textFlow.getChildren()
+                .stream()
+                .filter(Text.class::isInstance)
+                .map( t -> ((Text)t).getFont())
+                .filter(Objects::nonNull)
```

Comment:
I forgot, we need `.distinct()`.

Reply:
Good catch! Nice optimization. I think collecting to a `Set` will do same?

Reply:
Done